0000-lossy-conversions.md

```diff
@@ -0,0 +1,310 @@
+
+Truncation and Wrapping often occur together; for example, an `i32 → u16` conversion can both truncate and wrap around. To keep the complexity to a minimum, we treat wrapping as a special case of truncation, so we arrive at the following 3 new traits:
+
+- `TruncatingFrom<T>` — truncating conversions between integers
+- `SaturatingFrom<T>` — saturating conversions between integers
+- `ApproxFrom<T>` — lossy conversions that involve floats
+
+```rust
+pub trait TruncatingFrom<T> {
+    fn truncating_from(value: T) -> Self;
+}
+
+pub trait SaturatingFrom<T> {
+    fn saturating_from(value: T) -> Self;
+}
+
+pub trait ApproxFrom<T> {
+    fn approx_from(value: T) -> Self;
+}
+```
+
+`TruncatingFrom` and `ApproxFrom` can be implemented in the standard library using `as` by silencing the lint. For example:
+
+```rust
+#![allow(cast_lossy)]
+
+impl TruncatingFrom<i16> for i8 {
+    fn truncating_from(value: i16) -> i8 {
+        value as i8
+    }
+}
+
+impl ApproxFrom<f64> for f32 {
+    fn approx_from(value: f64) -> f32 {
+        value as f32
+    }
+}
+```
+
+`SaturatingFrom` must be implemented manually, but is straightforward:
+
+```rust
+#![allow(cast_lossy)]
+
+impl SaturatingFrom<i16> for i8 {
+    fn saturating_from(value: u8) -> i8 {
+        if value < i8::MIN as i16 {
+            i8::MIN
+        } else if value > i8::MAX as i16 {
+            i8::MAX
+        } else {
+            value as i8
+        }
+    }
+}
+```
+
+These traits are **unstable** for now. Before stabilizing them, we should consider adding `*Into` traits as well, but that discussion is left for the future.
+
+## Inherent methods
+
+[inherent-methods]: #inherent-methods
+
+Inherent methods similar to [`str::parse`](https://doc.rust-lang.org/std/primitive.str.html#method.parse) are added to make usage more ergonomic, e.g.
+
+```rust
+impl i32 {
+    pub fn truncate<T: TruncatingFrom<i32>>(self) -> T {
+        T::truncating_from(self)
+    }
+
+    pub fn saturate<T: SaturatingFrom<i32>>(self) -> T {
+        T::saturating_from(self)
+    }
+
+    pub fn approx<T: ApproxFrom<i32>>(self) -> T {
+        T::approx_from(self)
+    }
+}
+```
+
+This has several benefits. Unlike `value.truncating_into()` it allows specifying the output type, and unlike `T::truncating_from(value)`, it is chainable. Furthermore, inherent methods are always in scope and don't require importing a trait.
+
+## List of conversions
+
+This list of conversions should be implemented:
+
+- `TruncatingFrom` and `SaturatingFrom`:
+  - all **signed** to **unsigned** integers
+  - all **signed** to **smaller signed** integers (e.g. `i16 → i8`)
+  - all **unsigned** to **smaller or equal-sized** integers (e.g. `u32 → u16` or `u32 → i32`)
+  - specifically, for `isize` and `usize` (we assume they have 16 to 128 bits):
+    - `u32`, `u64`, `u128`, `i8`, `i16`, `i32`, `i64`, `i128` into `usize`
+    - `u16`, `u32`, `u64`, `u128`, `i32`, `i64`, `i128` into `isize`
```

Comment:
I understand that `u8` is not part of `isize` because `isize` is always a superset, but would it be more consistent to make them symmetric?

Reply:
Are you suggesting to implement saturating and truncating methods for _lossless_ conversions? I don't think that makes sense. Or what kind of symmetry are you looking for?